File: project/jimmer-apt/src/main/java/org/babyfish/jimmer/apt/dto/DtoProcessor.java
package org.babyfish.jimmer.apt.dto;

import org.babyfish.jimmer.Input;
import org.babyfish.jimmer.View;
import org.babyfish.jimmer.apt.Context;
import org.babyfish.jimmer.apt.client.DocMetadata;
import org.babyfish.jimmer.apt.immutable.generator.Constants;
import org.babyfish.jimmer.apt.immutable.meta.ImmutableProp;
import org.babyfish.jimmer.apt.immutable.meta.ImmutableType;
import org.babyfish.jimmer.apt.util.GenericParser;
import org.babyfish.jimmer.dto.compiler.*;

import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

public class DtoProcessor {

    private final Context context;

    private final Elements elements;

    private final Collection<String> dtoDirs;

    private final boolean dtoBundleEnabled;

    private final DtoModifier defaultNullableInputModifier;

    public DtoProcessor(
            Context context,
            Elements elements,
            Collection<String> dtoDirs,
            boolean dtoBundleEnabled,
            DtoModifier defaultNullableInputModifier
    ) {
        this.context = context;
        this.elements = elements;
        this.dtoDirs = dtoDirs;
        this.dtoBundleEnabled = dtoBundleEnabled;
        this.defaultNullableInputModifier = defaultNullableInputModifier;
    }

    public boolean process() {
        return generateDtoTypes(parseDtoTypes());
    }

    private List<DtoType<ImmutableType, ImmutableProp>> parseDtoTypes() {
        List<AptDtoCompiler> compilers = new ArrayList<>();
        DtoContext dtoContext = new DtoContext(context.getFiler(), dtoDirs, dtoBundleEnabled);
        AptDtoCompiler compiler;

        for (DtoFile dtoFile : dtoContext.getDtoFiles()) {
            try {
                compiler = new AptDtoCompiler(dtoFile, context, elements, defaultNullableInputModifier);
            } catch (DtoAstException ex) {
                throw new DtoException(
                        "Failed to parse \"" +
                                dtoFile.getAbsolutePath() +
                                "\": " +
                                ex.getMessage(),
                        ex
                );
            } catch (Throwable ex) {
                throw new DtoException(
                        "Failed to read \"" +
                                dtoFile.getAbsolutePath() +
                                "\": " +
                                ex.getMessage(),
                        ex
                );
            }
            compilers.add(compiler);
        }
        List<DtoType<ImmutableType, ImmutableProp>> dtoTypes = DtoCompiler
                .compileAll(compilers, context::includeDtoTarget)
                .values()
                .stream()
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
        DtoTypeLinker.link(dtoTypes, this::resolveDtoType);
        return dtoTypes;
    }

    private DtoTypeInfo<ImmutableType> resolveDtoType(String qualifiedName) {
        TypeElement typeElement = elements.getTypeElement(qualifiedName);
        if (typeElement == null) {
            return null;
        }
        Types types = context.getTypes();
        TypeMirror type = types.erasure(typeElement.asType());
        DtoTypeKind kind;
        String superName;
        if (types.isSubtype(type, types.erasure(elements.getTypeElement(Input.class.getName()).asType()))) {
            kind = DtoTypeKind.INPUT;
            superName = Input.class.getName();
        } else if (types.isSubtype(type, types.erasure(elements.getTypeElement(View.class.getName()).asType()))) {
            kind = DtoTypeKind.VIEW;
            superName = View.class.getName();
        } else if (types.isSubtype(
                type,
                types.erasure(
                        elements.getTypeElement(Constants.JSPECIFICATION_CLASS_NAME.canonicalName()).asType()
                )
        )) {
            kind = DtoTypeKind.SPECIFICATION;
            superName = Constants.JSPECIFICATION_CLASS_NAME.canonicalName();
        } else {
            return null;
        }
        TypeMirror baseTypeMirror = new GenericParser(
                "reusable DTO",
                typeElement,
                superName
        ).parse().arguments.get(0);
        ImmutableType baseType = context.getImmutableType(baseTypeMirror);
        if (baseType == null) {
            throw new DtoException(
                    "The entity type argument of reusable DTO type \"" +
                            qualifiedName +
                            "\" is not an immutable type"
            );
        }
        return new DtoTypeInfo<>(baseType, kind);
    }

    private boolean generateDtoTypes(List<DtoType<ImmutableType, ImmutableProp>> dtoTypes) {
        boolean result = false;
        DocMetadata docMetadata = new DocMetadata(context);
        for (DtoType<ImmutableType, ImmutableProp> dtoType : dtoTypes) {
            new DtoGenerator(context, docMetadata, dtoType).generate();
            result = true;
        }
        return result;
    }
}

File: project/jimmer-ksp/src/main/kotlin/org/babyfish/jimmer/ksp/dto/DtoProcessor.kt
package org.babyfish.jimmer.ksp.dto

import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.symbol.KSClassDeclaration
import org.babyfish.jimmer.Input
import org.babyfish.jimmer.View
import org.babyfish.jimmer.dto.compiler.*
import org.babyfish.jimmer.dto.compiler.Anno.EnumValue
import org.babyfish.jimmer.ksp.Context
import org.babyfish.jimmer.ksp.KspDtoCompiler
import org.babyfish.jimmer.ksp.client.DocMetadata
import org.babyfish.jimmer.ksp.immutable.meta.ImmutableProp
import org.babyfish.jimmer.ksp.immutable.meta.ImmutableType
import org.babyfish.jimmer.ksp.util.GenericParser
import org.babyfish.jimmer.ksp.util.fastResolve

class DtoProcessor(
    private val ctx: Context,
    private val mutable: Boolean,
    private val dtoDirs: Collection<String>,
    private val defaultNullableInputModifier: DtoModifier
) {
    fun process(): Boolean {
        val dtoTypeMap = findDtoTypeMap()
        generateDtoTypes(dtoTypeMap)
        return dtoTypeMap.isNotEmpty()
    }

    private fun findDtoTypeMap(): Map<ImmutableType, MutableList<DtoType<ImmutableType, ImmutableProp>>> {
        val dtoTypeMap = mutableMapOf<ImmutableType, MutableList<DtoType<ImmutableType, ImmutableProp>>>()
        val dtoCtx = DtoContext(ctx.resolver.getAllFiles().firstOrNull(), dtoDirs)
        val compilers = mutableListOf<KspDtoCompiler>()
        for (dtoFile in dtoCtx.dtoFiles) {
            val compiler = try {
                KspDtoCompiler(dtoFile, ctx, defaultNullableInputModifier)
            } catch (ex: DtoAstException) {
                throw DtoException(
                    "Failed to parse \"" +
                            dtoFile.absolutePath +
                            "\": " +
                            ex.message,
                    ex
                )
            } catch (ex: Throwable) {
                throw DtoException(
                    "Failed to read \"" +
                            dtoFile.absolutePath +
                            "\": " +
                            ex.message,
                    ex
                )
            }
            compilers += compiler
        }
        for (dtoTypes in DtoCompiler.compileAll(compilers, ctx::includeDtoTarget).values) {
            for (dtoType in dtoTypes) {
                dtoTypeMap.computeIfAbsent(dtoType.baseType) {
                    mutableListOf()
                } += dtoType
            }
        }
        DtoTypeLinker.link(dtoTypeMap.values.flatten(), ::resolveDtoType)
        ctx.resolve()
        return dtoTypeMap
    }

    private fun resolveDtoType(qualifiedName: String): DtoTypeInfo<ImmutableType>? {
        val declaration = ctx.resolver.getClassDeclarationByName(qualifiedName) ?: return null
        val inputType = ctx.resolver
            .getClassDeclarationByName(Input::class.qualifiedName!!)!!
            .asStarProjectedType()
        val viewType = ctx.resolver
            .getClassDeclarationByName(View::class.qualifiedName!!)!!
            .asStarProjectedType()
        val kind: DtoTypeKind
        val superName: String
        val type = declaration.asStarProjectedType()
        if (inputType.isAssignableFrom(type)) {
            kind = DtoTypeKind.INPUT
            superName = Input::class.qualifiedName!!
        } else if (viewType.isAssignableFrom(type)) {
            kind = DtoTypeKind.VIEW
            superName = View::class.qualifiedName!!
        } else {
            return null
        }
        val baseDeclaration = GenericParser(
            "reusable DTO",
            declaration,
            superName
        ).parse().arguments[0].type!!.fastResolve().declaration as? KSClassDeclaration
            ?: throw DtoException(
                "The entity type argument of reusable DTO type \"$qualifiedName\" " +
                    "is not an immutable type"
            )
        if (ctx.typeAnnotationOf(baseDeclaration) == null) {
            throw DtoException(
                "The entity type argument of reusable DTO type \"$qualifiedName\" " +
                    "is not an immutable type"
            )
        }
        return DtoTypeInfo(ctx.typeOf(baseDeclaration), kind)
    }

    private fun generateDtoTypes(
        dtoTypeMap: Map<ImmutableType, List<DtoType<ImmutableType, ImmutableProp>>>
    ) {
        val allFiles = ctx.resolver.getAllFiles().toList()
        val docMetadata = DocMetadata(ctx)
        for (dtoTypes in dtoTypeMap.values) {
            for (dtoType in dtoTypes) {
                val mutable = dtoType.annotations.firstOrNull {
                    it.qualifiedName == "org.babyfish.jimmer.kt.dto.KotlinDto"
                }?.let {
                    val value = it.valueMap["immutability"] as EnumValue
                    when (value.constant) {
                        "IMMUTABLE" -> false
                        "MUTABLE" -> true
                        else -> null
                    }
                } ?: mutable
                DtoGenerator(ctx, docMetadata, mutable, dtoType, ctx.environment.codeGenerator).generate(allFiles)
            }
        }
    }
}
